starknet_py/contract_abi_resolver.py

```diff
@@ -0,0 +1,131 @@
+    that will be used for checking if contract at the address is a proxy contract.
+    """
+
+
+def prepare_proxy_config(proxy_config: Union[bool, ProxyConfig]) -> ProxyConfig:
+    if isinstance(proxy_config, bool):
+        if not proxy_config:
+            return ProxyConfig()
+        proxy_config = ProxyConfig()
+
+    if "max_steps" in proxy_config:
+        warnings.warn(
+            "ProxyConfig.max_steps is deprecated. Contract.from_address always makes at most 1 step.",
+            category=DeprecationWarning,
+        )
+    proxy_checks = [OpenZeppelinProxyCheck(), ArgentProxyCheck()]
+    return {"proxy_checks": proxy_checks, **proxy_config}
+
+
+class ImplementationType(Enum):
+    """
+    Enum representing transaction types
+    """
+
+    CLASS_HASH = "class_hash"
+    ADDRESS = "address"
+
+
+class ContractAbiResolver:
+    """
+    Class for resolving abi of a contract
+    """
+
+    @staticmethod
+    async def resolve_direct(
+        address: Address,
+        client: Client,
+    ) -> Abi:
+        """
+        Returns abi of a contract directly from given address.
+        """
+        contract_class = await ContractAbiResolver._get_class_by_address(
+            address=address, client=client
+        )
+        return contract_class.abi or []
+
+    @staticmethod
+    async def resolve_proxy_step(
+        proxy_contract: "Contract",
+        proxy_config: ProxyConfig,
+    ) -> Abi:
+        """
+        Returns abi of a contract that is being proxied by proxy_contract.
+        """
+        impl = await ContractAbiResolver._get_implementation_from_proxy(
+            proxy_contract=proxy_contract, proxy_config=proxy_config
+        )
+        implementation, implementation_type = impl
+        proxy_client = proxy_contract.client
+
+        if implementation_type == ImplementationType.CLASS_HASH:
+            contract_class = await proxy_client.get_class_by_hash(implementation)
+        elif implementation_type == ImplementationType.ADDRESS:
+            contract_class = await ContractAbiResolver._get_class_by_address(
+                address=implementation, client=proxy_client
+            )
+        else:
+            return []
+
+        return contract_class.abi or []
+
+    @staticmethod
+    async def _get_implementation_from_proxy(
+        proxy_contract: "Contract", proxy_config: ProxyConfig
+    ) -> Tuple[int, ImplementationType]:
+        implementation, proxy_checks = None, proxy_config.get("proxy_checks", [])
+        for proxy_check in proxy_checks:
+            implementation = await proxy_check.implementation_hash(proxy_contract)
+            if implementation is not None:
+                return implementation, ImplementationType.CLASS_HASH
+
+            implementation = await proxy_check.implementation_address(proxy_contract)
+            if implementation is not None:
+                return implementation, ImplementationType.ADDRESS
+
+        raise ProxyResolutionError(proxy_checks)
```

Comment:
Looks like this exception is not a part of Proxies themselves, but `ContractAbiResolver` itself I suggest moving this exception to this file.